include/revng/Model/Type.h

```diff
@@ -0,0 +1,639 @@
+  revng_assert(QT.verify());
+  return not QT.Qualifiers.empty();
+}
+
+RecursiveCoroutine<std::size_t> typeSize(const model::Type *T);
+
+RecursiveCoroutine<std::size_t> typeSize(const model::QualifiedType &T);
+
+} // end namespace model
+
+/// \brief A primitive type in model: sized integers, booleans, floats and void.
+struct model::PrimitiveType : public model::Type {
+  static constexpr const char *Tag = "!model::PrimitiveType";
+
+  PrimitiveType(llvm::StringRef NameRef) : Type(NameRef, TypeKind::Primitive) {}
+  PrimitiveType() : PrimitiveType("") {}
+
+  static bool classof(const Type *T) { return T->Kind == TypeKind::Primitive; }
+
+private:
+  // Kind is TypeKind::Primitive. Name is allowed by PrimitiveTypeNames.
+  bool verifyDerived() const debug_function;
+
+  friend struct model::Type;
+};
+// Enable introspection for model::PrimitiveType
+INTROSPECTION_NS(model, PrimitiveType, ID, Name, Kind);
+
+/// \brief KeyedObjectTraits for std::string based on its value
+template<>
+struct KeyedObjectTraits<std::string>
+  : public IdentityKeyedObjectTraits<std::string> {};
+
+/// \brief An entry in a model enum, with a name and a value.
+struct model::EnumEntry {
+
+  std::size_t Value;
+  std::string Name;
+  SortedVector<std::string> Aliases;
+
+public:
+  EnumEntry(std::size_t V, llvm::StringRef NameRef) :
+    Value(V), Name(NameRef), Aliases() {}
+
+  EnumEntry(std::size_t V) : EnumEntry(V, "") {}
+
+  EnumEntry() : EnumEntry(0ULL) {}
+
+  bool operator==(const EnumEntry &) const = default;
+
+public:
+  // The entry should have a non-empty name, the name should not be a valid
+  // alias, and there should not be empty aliases.
+  bool verify() const debug_function;
+};
+INTROSPECTION_NS(model, EnumEntry, Value, Name, Aliases);
+
+/// \brief KeyedObjectTraits for model::EnumEntry based on its value
+template<>
+struct KeyedObjectTraits<model::EnumEntry> {
+
+  static std::size_t key(const model::EnumEntry &E) { return E.Value; }
+
+  static model::EnumEntry fromKey(std::size_t V) {
+    return model::EnumEntry{ V };
+  }
+};
+
+/// \brief Make EnumEntry yaml-serializable
+template<>
+struct llvm::yaml::MappingTraits<model::EnumEntry>
+  : public TupleLikeMappingTraits<model::EnumEntry> {};
+
+/// \brief An enum type in model. Enums are actually typedefs of unnamed
+/// enums.
+struct model::EnumType : public model::Type {
+  static constexpr const char *Tag = "!model::EnumType";
+
+  Path UnderlyingType;
+  SortedVector<EnumEntry> Entries;
+
+public:
+  EnumType(llvm::StringRef NameRef) :
+    Type(NameRef, TypeKind::Enum), UnderlyingType(), Entries() {}
+
+  EnumType() : EnumType("") {}
+
+  static bool classof(const Type *T) { return T->Kind == TypeKind::Enum; }
+
+private:
+  // Kind must be TypeKind::Enum; Underlying type must verify and be an
+  // integral primitive type; there must be at least one entry, and all the
+  // entries must verify.
+  bool verifyDerived() const debug_function;
+
+  friend struct model::Type;
```

Comment:
Why?

Reply:
So that `Type::verify()` can do the same baseline checks for all types, then call `llvm::cast<Derived>->verifyDerived()` for stuff that depends on the single field. 
For now I switched to `friend bool model::Type::verify() cons`, so that only verification is friend.

Any better idea?

Reply:
Idea:
* `Type::verify()` verifies itself only
* `Derived::verify()` calls `Type::verify()` and verifies `Derived`-specific stuff.
* If you have `Upcastable<Type> A` you do `upcast(A, [] (auto &Upcasted) { Upcasted.verify() })`.

Upcastable pointers are designed to be used with templated lambdas magic.

Reply:
I don't like the idea. Pretty sure that someone (probably me) is going to forget to `upcast` in a place where calling only `Type::verify()` will give a fake sense of safety, while getting away with a broken derived thing.

I could change `verify` to be `virtual`, but that would be the only reason to have a vtable here.

But I really want to be impossible to use `verify` in the wrong way.

So here's another idea:
- Rename `Type::verify()` to `Type::verifyBase()` and make it `protected`. In this way, `Type` does not have `verify` and it cannot be called wrongly.
- `Derived::verify()` calls `Type::verifyBase()` and verifies `Derived` stuff
- In this way, users cannot do `MyBaseType.verify()` not `MyBaseTypePtr->verify()` because `verify` is not there on the base type. The only thing they can do is to `upcast(A, [](auto &Upcasted){ Upcasted.verify() }), for which we could provide a helper `bool verify(const Upcastable<Type> &U)` that does the upcast internally.

Reply:
Works, but at this point you could keep the `Type::verify` as a name and just make it protected.

Reply:
Works. I'll do this.

Reply:
None of the ideas suggested above works, because they all require you to hold an `UpcastablePointer` in order to `verify()` a `Type`. In particular, in our case this not always true, e.g. when you are traversing a `TupleTreeReference` from one `Type` to another, calling `get()` on the `TupleTreeReference` gives you a `Type *`, not an `UpcastablePointer<Type>`. We could work around this, making `TupleTreeReference` return you an `UpcastablePointer`, not just the naked pointer it wraps. But I'm not sure I like this idea more than the original implementation with `verifyBase()`, `verifyDerived()` and the `friend bool model::Type::verify()` declaration in derived classes.

What do you think?

Reply:
No, ok, I managed to do it. Take a look.

Reply:
Looks amazing!